### DISCUSSION TITLE
Building dagger/dagger in a git worktree fails

### DISCUSSION THREAD

**Discussion by jedevc** [dagger/dagger#12209] [maintainers]
This is an issue we can easily encounter on the `main` branch of https://github.com/dagger/dagger.

Create a worktree:

```
$ git worktree add <branch-name>
$ cd <branch-name>
```

Try and get some git metadata:

```
$ dagger call -m version git head
...
fatal: not a git repository: /home/jedevc/Documents/work/dagger/dagger/.git/worktrees/<branch-name>
```

This occurs because we look for where `.git` exists - there's a git file created at `./<branch-name>/.git` - it's not a directory, and instead contains contents like: `gitdir: /home/jedevc/Documents/work/dagger/dagger/.git/worktrees/<branch-name>`. While the whole module loads correctly, we encounter an issue when trying to load `.git` as a contextual directory - which makes sense.

A couple possible solutions:
- On `findUp`, skip the `.git` worktree, and go all the way to the top `.git` directory. This kind of works, but now we get all sorts of breakages from having a different root - e.g. contextual directories would now resolve to `/` which is the top-level git repo, instead of the worktree repo (which is probably what the user wants).
	- This is also a tangled mess, because the module loading code is *so* nested and complex.
- Avoid the need to transfer a contextual `.git` directory in the first place - instead use a first-class contextual git feature, as suggested here: https://github.com/dagger/dagger/issues/8520#issuecomment-2368105528.

Not entirely sure, but recording this so I don't forget about it - it would be nice to fix.

**Comment by eternalphane** [comment]
Got (almost) the same problem when trying to call `Directory.asGit` in a git submodule.
